storagenode/contact/chore.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package contact
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/satellite/overlay"
+	"storj.io/storj/storagenode/trust"
+)
+
+var mon = monkit.Package()
+
+// Config contains configurable parameters for contact chore
+type Config struct {
+	Interval time.Duration `help:"how frequently the node contact chore should run" releaseDefault:"1h" devDefault:"30s"`
+	MaxSleep time.Duration `help:"maximum duration to wait before pinging satellites" releaseDefault:"45m" devDefault:"0s" hidden:"true"`
+}
+
+// Chore is the contact chore for nodes announcing themselves to their trusted satellites
+type Chore struct {
+	log       *zap.Logger
+	self      overlay.NodeDossier
+	transport transport.Client
+
+	trust *trust.Pool
+
+	maxSleep time.Duration
+	Loop     *sync2.Cycle
+}
+
+// NewChore creates a new contact chore
+func NewChore(log *zap.Logger, interval time.Duration, maxSleep time.Duration, trust *trust.Pool, transport transport.Client, self overlay.NodeDossier) *Chore {
+	return &Chore{
+		log:       log,
+		self:      self,
+		transport: transport,
+
+		trust: trust,
+
+		maxSleep: maxSleep,
+		Loop:     sync2.NewCycle(interval),
+	}
+}
+
+// Run the contact chore on a regular interval with jitter
+func (chore *Chore) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	chore.log.Info("Storagenode contact chore starting up")
+
+	return chore.Loop.Run(ctx, func(ctx context.Context) error {
+		if err := chore.sleep(ctx); err != nil {
+			return err
+		}
+		if err = chore.pingSatellites(ctx); err != nil {
```

Comment:
This assignment should probably be `err := ` instead of `err = ` to make it more obvious that this error value does not need to live past this block. Written as it is, this `err` becomes a capture of the one declared as a named return on line 57, which is weird because `Chore.Run()` will probably return before most invocations of this anonymous function are even called. All invocations will use the same memory location for this variable, and it won't necessarily be initialized to `nil` each time, and oh man if two of these get invoked at the same time this would probably even cause a race condition. None of this ought to be a problem with the way the code is written right now, but if the code changes this could be a source of some really impressively subtle bugs 😀